hail/src/main/scala/is/hail/types/physical/PCanonicalBaseStruct.scala

```diff
@@ -192,10 +192,11 @@ abstract class PCanonicalBaseStruct(val types: Array[PType]) extends PBaseStruct
         addrVar := addr,
         stagedInitialize(addrVar, setMissing = true),
         Code(fields.zip(srcStruct.fields).map { case (dest, src) =>
-          assert(dest.typ.required <= src.typ.required)
           val idx = dest.index
           assert(idx == src.index)
-          srcStruct.isFieldDefined(srcAddrVar, idx).orEmpty(Code(
+          srcStruct.isFieldMissing(srcAddrVar, idx).mux(
```

Comment:
why do we need this change? This relaxes type copying to do runtime checks, which maybe means our ptype selection is wrong.

Reply:
I think this may have been me attempting to fix this:

https://github.com/hail-is/hail/pull/8800/commits/abdd76459d031e6a8e2bdba415dab6edd1ca909f#diff-f5c3e98d14c9ed442da71493a0d75913L1250

and fixing it in the wrong place. I've backed it out in favor of the other thing.